### PR TITLE
add support for forcing commit or waiting for a commit on file ingest…

### PR DESCRIPTION
**Pull request by trinity-1686a**
### Description

add command line arguments to wait for/force a commit when ingesting from the cli

### How was this PR tested?

setting a very high commit timeout, and verifying a commit get trigger at the end of the ingestion

in practice it almost always hit [this timeout](https://github.com/quickwit-oss/quickwit/pull/3049/files#diff-4c0fcf5f3d94686949a73e623bde2f115360207eb410946a969fefae1605feceR83). It should be made much bigger for `commit=wait_for/force`, but maybe not for `commit=auto`

### CODE REVIEW

File: quickwit/quickwit-cli/src/index.rs [hunk 1]
```diff
@@ -121,6 +121,16 @@ pub fn build_index_command<'a>() -> Command<'a> {
                         .display_order(1),
                     arg!(--"input-path" <INPUT_PATH> "Location of the input file.")
                         .required(false),
+                    Arg::new("wait")
+                        .long("wait")
+                        .short('w')
+                        .help("Enable waiting after the last document is sent")
```
[the remark this answers is not in the document]
Comment (fmassot) [reply quickwit/quickwit-cli/src/index.rs:127]: `after the last document is sent`

sent to who?

File: quickwit/quickwit-rest-client/src/rest_client.rs [hunk 2]
```diff
@@ -154,10 +156,24 @@ impl QuickwitClient {
...
-                    .send::<()>(Method::POST, &ingest_path, None, None, Some(batch.clone()))
-                    .await?;
+                let response = if let (Some(query_param), false) = (
+                    last_block_commit.to_query_parameter(),
+                    batch_reader.has_next(),
+                ) {
+                    self.transport
+                        .send::<()>(
+                            Method::POST,
+                            &format!("{ingest_path}?{query_param}"),
```
Comment (fmassot) [inline quickwit/quickwit-rest-client/src/rest_client.rs:166]: you should use the `query_string` argument that is used for passing your query parameters.

Comment (fmassot) [reply quickwit/quickwit-rest-client/src/rest_client.rs:166]: I would simplify the code like this:

```rust

let query_params = if !batch_reader.has_next() {
    last_block_commit.to_query_parameter()
};
self.transport.send::<()>...
```

**Review by fmassot — CHANGES_REQUESTED** [review]
See comments.

File: quickwit/quickwit-cli/src/index.rs [hunk 3]
```diff
@@ -323,11 +334,18 @@ impl IndexCliCommand {
         } else {
             None
         };
+        let commit_type = match (matches.is_present("wait"), matches.is_present("force")) {
+            (false, false) => CommitType::Auto,
+            (false, true) => CommitType::Force,
+            (true, false) => CommitType::WaitFor,
+            (true, true) => bail!("Can't specify both --wait and --force"),
```
Comment (imotov) [inline quickwit/quickwit-cli/src/index.rs:341]: You can probably bail out even earlier during command line parsing by using ArgGroup or the conflicts_with method.

File: quickwit/quickwit-cli/src/index.rs [hunk 4]
```diff
@@ -121,6 +121,16 @@ pub fn build_index_command<'a>() -> Command<'a> {
                         .display_order(1),
                     arg!(--"input-path" <INPUT_PATH> "Location of the input file.")
                         .required(false),
+                    Arg::new("wait")
+                        .long("wait")
+                        .short('w')
+                        .help("Wait for a document to be commited before exiting")
```
Comment (imotov) [inline quickwit/quickwit-cli/src/index.rs:127]: Maybe "Wait for all documents to be committed and available for search before exiting"?

File: quickwit/quickwit-cli/src/index.rs [hunk 5]
```diff
@@ -124,13 +124,14 @@ pub fn build_index_command<'a>() -> Command<'a> {
...
                         .short('w')
-                        .help("Wait for a document to be commited before exiting")
+                        .help("Wait for all documents to be commited and available for search before exiting")
                         .takes_value(false),
                     Arg::new("force")
                         .long("force")
                         .short('f')
-                        .help("Force a commit after the last document is sent, and wait for the commit to finish")
-                        .takes_value(false),
+                        .help("Force a commit after the last document is sent, and wait for all documents to be commited .nd available before exiting")
```
Comment (fmassot) [inline quickwit/quickwit-cli/src/index.rs:132]:
```suggestion
                        .help("Force a commit after the last document is sent, and wait for all documents to be committed and available for search before exiting")
```

**Review by fmassot — APPROVED** [review]
Just one little typo to fix and we are good with the PR!

**Pull request merged**